rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -0,0 +1,613 @@
+  *sequence_id = sequence_number_++;
+
+  // Send request
+  z_get_options_t opts = z_get_options_default();
+  z_owned_bytes_map_t map = create_map_and_set_sequence_num(
+    *sequence_id,
+    [this](z_owned_bytes_map_t * map, const char * key)
+    {
+      uint8_t local_gid[RMW_GID_STORAGE_SIZE];
+      entity_->copy_gid(local_gid);
+      z_bytes_t gid_bytes;
+      gid_bytes.len = RMW_GID_STORAGE_SIZE;
+      gid_bytes.start = local_gid;
+      z_bytes_map_insert_by_copy(map, z_bytes_new(key), gid_bytes);
+    });
+  if (!z_check(map)) {
+    // create_map_and_set_sequence_num already set the error
+    return RMW_RET_ERROR;
+  }
+  auto always_free_attachment_map = rcpputils::make_scope_exit(
+    [&map]() {
+      z_bytes_map_drop(z_move(map));
+    });
+
+  // See the comment about the "num_in_flight" class variable in the ClientData class for
+  // why we need to do this.
+  {
+    std::lock_guard<std::mutex> lock(num_in_flight_mutex);
+    auto num_in_flight_it = num_in_flight_map.find(this);
+    if (num_in_flight_it == num_in_flight_map.end()) {
+      num_in_flight_map[this] = 0;
+      num_in_flight_it = num_in_flight_map.find(this);
+    }
+    num_in_flight_it->second++;
+  }
+  opts.attachment = z_bytes_map_as_attachment(&map);
+  opts.target = Z_QUERY_TARGET_ALL_COMPLETE;
+  // The default timeout for a z_get query is 10 seconds and if a response is not received within
+  // this window, the queryable will return an invalid reply. However, it is common for actions,
+  // which are implemented using services, to take an extended duration to complete. Hence, we set
+  // the timeout_ms to the largest supported value to account for most realistic scenarios.
+  opts.timeout_ms = std::numeric_limits<uint64_t>::max();
+  // Latest consolidation guarantees unicity of replies for the same key expression,
+  // which optimizes bandwidth. The default is "None", which imples replies may come in any order
+  // and any number.
+  opts.consolidation = z_query_consolidation_latest();
+  opts.value.payload = z_bytes_t{data_length, reinterpret_cast<const uint8_t *>(request_bytes)};
+  // TODO(Yadunund): Once we switch to zenoh-cpp with lambda closures,
+  // capture shared_from_this() instead of this.
+  z_owned_closure_reply_t zn_closure_reply =
+    z_closure(client_data_handler, client_data_drop, this);
+  z_get(
+    context_impl->session(),
+    z_loan(keyexpr_), "",
+    z_move(zn_closure_reply),
+    &opts);
+
+  return RMW_RET_OK;
+}
+
+///=============================================================================
+ClientData::~ClientData()
+{
+  const rmw_ret_t ret = this->shutdown();
+  if (ret != RMW_RET_OK) {
+    RMW_ZENOH_LOG_ERROR_NAMED(
+      "rmw_zenoh_cpp",
+      "Error destructing client /%s.",
+      entity_->topic_info().value().name_.c_str()
+    );
+  }
+  std::lock_guard<std::mutex> lock(num_in_flight_mutex);
+  num_in_flight_map.erase(this);
+  deleted_clients.insert(this);
```

Comment:
Importantly, I think we only need to add this to the `deleted_clients` map if there is anything in flight.  I think that should keep us from "leaking" memory in the `deleted_clients` map for every deleted client.